### PR TITLE
Added option to update parameters using state_dict in AveragedModel

### PR DESCRIPTION
**Pull request by prabhat00155**
While implementing [EMA](https://github.com/pytorch/vision/pull/4381)(which extends AveragedModel) in torchvision, update_parameters() from AveragedModel could not be used as it did not handle state_dict(), so a custom update_parameters() needed to be defined in [EMA class](https://github.com/pytorch/vision/pull/4406). This PR aims to handle this scenario removing the need for this custom update_parameters() implementation.

Discussion: https://github.com/pytorch/vision/pull/4406#pullrequestreview-753734102

### CODE REVIEW

File: torch/optim/swa_utils.py [hunk 1]
```diff
@@ -100,8 +100,10 @@ def avg_fn(averaged_model_parameter, model_parameter, num_averaged):
     def forward(self, *args, **kwargs):
         return self.module(*args, **kwargs)
 
-    def update_parameters(self, model):
-        for p_swa, p_model in zip(self.parameters(), model.parameters()):
+    def update_parameters(self, model, use_state_dict=False):
```
Comment (datumbox) [inline torch/optim/swa_utils.py:103]: I seems more natural to define this on the constructor as one won't switch between different options in the middle of the updates.

**Review by datumbox — COMMENTED** [review]
Thanks for the PR @prabhat00155. 

I agree that this is very important for models with layers contain buffers such as BatchNorm. Not averaging all the parameters leads to lower performance. 

Due to this missing feature, the community has build multiple implementations to get around the problem [[1](https://github.com/facebookresearch/ClassyVision/blob/c845f0430fed099f0ee74370d0417316ada253e5/classy_vision/hooks/exponential_moving_average_model_hook.py#L61-L73), [2](https://github.com/pytorch/vision/blob/main/references/classification/utils.py#L175-L184), [3](https://github.com/rwightman/pytorch-image-models/blob/a2727c1bf78ba0d7b5727f5f95e37fb7f8866b1f/timm/utils/model_ema.py#L117)]. So this is definitely worth fixing.

Below I left a comment for your attention, please let me know what you think. Also we probably need to add a unit-test that covers the new code.

File: torch/optim/swa_utils.py (hunk 1 above)
Comment (prabhat00155) [reply torch/optim/swa_utils.py]: Makes sense.

File: test/test_optim.py [hunk 2]
```diff
@@ -2290,6 +2290,34 @@ def avg_fn(p_avg, p, n_avg):
         for p_avg, p_swa in zip(averaged_params, averaged_dnn.parameters()):
             self.assertEqual(p_avg, p_swa)
 
+    def test_averaged_model_exponential_use_state_dict(self):
+        # Test AveragedModel with EMA as avg_fn and use_state_dict as True.
+        dnn = torch.nn.Sequential(
+            torch.nn.Conv2d(1, 5, kernel_size=3),
```
Comment (datumbox) [inline test/test_optim.py:2296]: Because your PR is specifically addressing issues with layers that have buffers it would be great to add a BatchNorm layer here.

**Review by datumbox — APPROVED** [review]
LGTM, thanks @prabhat00155.

I left one comment for your consideration but it's not blocking.

File: torch/optim/swa_utils.py [hunk 3]
```diff
@@ -84,7 +84,7 @@ class AveragedModel(Module):
         Generalizes Well:
         https://arxiv.org/abs/2001.02312
     """
-    def __init__(self, model, device=None, avg_fn=None):
+    def __init__(self, model, device=None, avg_fn=None, use_state_dict=False):
```
Comment (datumbox) [inline torch/optim/swa_utils.py:87]: Also the parameter `use_state_dict` needs to be added in the documentation of the class.

File: test/test_optim.py (hunk 2 above)
Comment (prabhat00155) [reply test/test_optim.py:2296]: Makes sense, done.

File: torch/optim/swa_utils.py (hunk 3 above)
Comment (prabhat00155) [reply torch/optim/swa_utils.py]: Added.

File: torch/optim/swa_utils.py [hunk 4]
```diff
@@ -96,12 +98,15 @@ def avg_fn(averaged_model_parameter, model_parameter, num_averaged):
                 return averaged_model_parameter + \
                     (model_parameter - averaged_model_parameter) / (num_averaged + 1)
         self.avg_fn = avg_fn
+        self.use_state_dict = mode == 'state_dict'
```
Comment (albanD) [inline torch/optim/swa_utils.py:101]: That looks very britle if the user pass another string as input.

File: torch/optim/swa_utils.py [hunk 5]
```diff
@@ -26,6 +26,8 @@ class AveragedModel(Module):
             :class:`AveragedModel` parameter, the current value of :attr:`model`
             parameter and the number of models already averaged; if None,
             equally weighted average is used (default: None)
+        mode (str, optional): whether to use parameters or state_dict for update
```
Comment (albanD) [inline torch/optim/swa_utils.py:29]: This should document what are the valid values for this argument.

**Review by albanD — COMMENTED** [review]
Is there a design discussion for this?
There can be objects that are not Tensors in the state_dict so I am not convinced this is going to work well.

**Comment by datumbox** [comment]
@albanD Thanks for the feedback. The PR contains links to the original issues/discussions at Vision but let me summarize the details here and make the post self-contained.

## Previous implementation shortcomings 

PyTorch's `AveragedModel` class didn't handle properly the buffers during model updates. This caused issues with many normalization layers (such as `BatchNorm2d`) that led to a significant decrease in accuracy (happy to provide additional numerical evidence; this is a [very well-known fact](https://www.internalfb.com/code/fbsource/[9f9a04823b76e3544a62b07d81e1bfd1e346afcf]/fbcode/stl/lightning/callbacks/ema.py?lines=29-35) in the research community). Another unfortunate detail of the previous implementation was that the initial values of the buffers were copied due to the `deepcopy` call, but were not subsequently updated by the `update_parameters()` method, leading to discrepancies:
https://github.com/pytorch/pytorch/blob/27135f86fd2f8b6d1d9e4b09719373761ffd49a7/torch/optim/swa_utils.py#L87-L89

## Duplicate implementations

Due to the above, an overwhelming number of libraries on the PyTorch ecosystem (both FB-backed and community led) are currently not using PyTorch's `AveragedModel` class. Instead they reimplement the functionality on their own. Here are a few libraries that do this along with links to their implementations:
- [TorchVision](https://github.com/pytorch/vision/blob/c8e3b2a5925e7b7ed21662e86a7e9553170a5633/references/classification/utils.py#L175-L184)
- [FairSeq](https://github.com/pytorch/fairseq/blob/fcca32258c8e8bcc9f9890bf4714fa2f96b6b3e1/fairseq/models/ema/ema.py#L139)
- [STL](https://www.internalfb.com/code/fbsource/[9f9a04823b76e3544a62b07d81e1bfd1e346afcf]/fbcode/stl/lightning/callbacks/ema.py?lines=29-35)
- [ClassyVision](https://github.com/facebookresearch/ClassyVision/blob/c845f0430fed099f0ee74370d0417316ada253e5/classy_vision/hooks/exponential_moving_average_model_hook.py#L61-L73)
- FAIR's [pycls](https://github.com/facebookresearch/pycls/blob/f8cd962737e33ce9e19b3083a33551da95c2d9c0/pycls/core/net.py#L129-L131)
- [D2go](https://github.com/facebookresearch/d2go/blob/82295dbf056c74501bb4470202dd4ad1e4e8e277/d2go/modeling/model_ema.py#L99-L101)
- [Timm](https://github.com/rwightman/pytorch-image-models/blob/a2727c1bf78ba0d7b5727f5f95e37fb7f8866b1f/timm/utils/model_ema.py#L117)
- [OpenMMLab](https://github.com/open-mmlab/mmdetection/blob/bde7b4b7eea9dd6ee91a486c6996b2d68662366d/mmdet/core/hook/ema.py#L57)

I strongly believe that improving the current implementation of the `AveragedModel` will bring significant benefits to PyTorch, the downstream libraries and the community. PyTorch's class will no longer be ignored, the downstream libraries won't have to reinvent the wheel and the community will get a utility that works off the shelf and can boost significant their accuracy.

## Different options

There are two generally accepted ways to handle the buffers:
1. Average their values across checkpoints (like we do for parameters)
2. Set their values directly to the latest checkpoint

There are pros/cons in both but the vast majority of the above implementations follow option 1. Implementing option 2 is also possible though less popular.

## The approach of this PR

@prabhat00155's PR introduces a new `mode` parameter on the API that will allow us to support multiple options on the future. Though many people consider the current behaviour of the class a bug, changing the original behaviour now would be too drastic and would cause BC issues. This is why we decided to set the default value of `mode` in a way that maintains the original functionality. 

Note that after merging this PR, the majority of the above implementations are now able to switch to using PyTorch core's `AveragedModel`. TorchVision plans to make the switch ASAP, when this change appears on the nightly.

## Is it safe to use state_dict()?

I believe it is safe. The vast majority of the linked implementations use `state_dict()` for this type of averaging. As a result this approach is very well-tested across a plethora of models and ML tasks.

Moreover the declared return type of the `state_dict()` method of `nn.Module`. Currently the PyTorch implementation indicates that it should return a dictionary with strings and Tensors:
https://github.com/pytorch/pytorch/blob/b80bdcc73b9d3af19e7e628e521cfe82b06ecb93/torch/nn/modules/module.py#L1278-L1287 

## Improving this implementation

I think your comments on the code of this implementations are fair and worth addressing. We will assess the options carefully and follow up with another PR. Shall we add you as a reviewer to ensure all is well prior merging? 

Finally please note that the TorchVision team plans to contribute more to Core on the future and we are happy to adjust to the processes that you use. Could you please clarify who is the POC for these utils? I ask because git blaming didn't show an obvious choice and we didn't know who to include from your side.

Let me know what you think. Thanks!

File: torch/optim/swa_utils.py (hunk 5 above)
Comment (prabhat00155) [reply torch/optim/swa_utils.py:29]: parameters and state_dict are valid values here as stated in line 29. Although, anything passed other than state_dict defaults to parameters.

Comment (albanD) [reply torch/optim/swa_utils.py:29]: Ho yes, what I meant is that this is missing `"` to make it clear that these are strings.
It's a nit though.

**Comment by albanD** [comment]
Thanks for the detailed comment @datumbox !

I wasn't trying to say that the change doesn't make sense. Just was wondering if there was an issues associated with this on pytorch/pytorch side.

> Moreover note that the declared return type of the Module.state_dict()

That's a great point. cc @jbschlosser should we change this signature now that we have `*_extra_state()` functions that can return anything?

I am still wondering if we wouldn't prefer to do always parameters and optionally buffers. That would avoid this issue altogether. And that would make the arg in the API a boolean which is easier to handle (not magic string matching).
But I guess that's too late.

> Could you please clarify who is the POC for these utils?

`torch.optim` has gone back into being maintained by the core team as a whole I'm afraid as our previous POC is not working on this anymore. You can have me as a POC if you need and I can route the review internally.

For these utils functions in optim, I am not sure we really want to grow them long term. These were adding as utilities for the swa optimizer but not really designed to be used in a standalone way.
If you plan to invest of some of the things here, we most likely want to move it somewhere else where there is better ownership. What do you think?

**Comment by datumbox** [comment]
@albanD No worries at all. That gave us the opportunity to document things properly on Core side. Sorry for bringing this PR without the necessary documentation in the first place. :) 

> I am still wondering if we wouldn't prefer to do always parameters and optionally buffers. 

If indeed the `state_dict()` is no longer guaranteed to return Tensors, we could make an internal change to handle params and buffers independently as you suggest. We will just need to come up with a better name for the `mode` value.

> make the arg in the API a boolean 

This will work only if we are certain we will never add option 2 listed above. I'm not saying that our team wants to add it now, just that the string arg gives us the option to do it on the future. 

> But I guess that's too late.

We can do a quick iteration and fix things if need be. We will just need to cherrypick them and put them on the release. So we kind of need to do it fast. If you can help with the reviews and clarify the design questions that popped up, we can do the coding part quickly.

> If you plan to invest of some of the things here, we most likely want to move it somewhere else where there is better ownership.

Makes sense, let's cross that bridge when we get there. For this round of review I'll have you as POC as you suggested.

**Comment by malfet** [comment]
> > make the arg in the API a boolean
> 
> This will work only if we are certain we will never add option 2 listed above. I'm not saying that our team wants to add it now, just that the string arg gives us the option to do it on the future.

Not sure if this would be something completely out of ordinary, but can we use `enum` here?

**Comment by datumbox** [comment]
> Option 2 seems orthogonal no?

Yes, option 1 and option 2 are not combined together. But you could support both strategies on the same class.

> Not sure if this would be something completely out of ordinary, but can we use enum here?

At TorchVision we started showing some love for enums. :) I haven't seen this pattern used much on core though.


Concerning next steps, we could:
1. Do better validation of `mode` and throw the right errors to the user: https://github.com/pytorch/pytorch/pull/65495#discussion_r717929493
2. Document better the available options: https://github.com/pytorch/pytorch/pull/65495#discussion_r718680583

Is there anything else we would like to do for this release?

**Comment by albanD** [comment]
That sounds good for me! Thanks for taking the time to do it.


> But you could support both strategies on the same class.

Yes, but it is not clear to me if it should be 2 boolean arguments or a single string/enum argument.
